chapters/categorical-predictors.qmd

```diff
@@ -0,0 +1,718 @@
+
+```{r}
+#| label: hotel-rate-hash-calcs
+#| warning: false
+#| error: false
+hash_10 <-
+  hotel_rate_train %>%
+  distinct(agent) %>%
+  recipe(~ agent) %>%
+  step_dummy_hash(agent, keep_original_cols = TRUE, num_terms = 10) %>%
+  prep() %>%
+  bake(new_data = NULL) %>%
+  rename_at(vars(-agent), ~ gsub("dummyhash_agent_", "", .x)) %>%
+  relocate(agent) %>%
+  arrange(agent)
+
+hash_256 <-
+  hotel_rate_train %>%
+  distinct(agent) %>%
+  recipe(~ agent) %>%
+  step_dummy_hash(agent, keep_original_cols = TRUE, num_terms = 2^8) %>%
+  prep() %>%
+  bake(new_data = NULL) %>%
+  rename_at(vars(-agent), ~ gsub("dummyhash_agent_", "hash ", .x)) %>%
+  relocate(agent) %>%
+  arrange(agent)
+```
+
+```{r}
+#| label: tbl-feature-hash
+#| tbl-cap: "Singed indicators for agent via feature hashing."
+
+remake_name <- function(x) {
+  x <- gsub("_", " ", x)
+  stringr::str_to_title(x)
+}
+
+hash_summary <-
+    hash_10 %>%
+      summarise(
+        across(c(-agent), 
+               ~ sprintf("%.3.1f", mean(abs(.x), na.rm = TRUE) * 100))
+      ) %>%
+      mutate(agent = c("density (%)"), .before = 1) %>% 
+  as.data.frame() %>% 
+  format() %>% 
+  as.data.frame()
+
+rownames(hash_summary) <- " "
+
+hash_10_chr <- 
+  hash_10 %>%
+  mutate(agent = gsub("_", " ", agent)) %>% 
+  as.data.frame() %>% 
+  format() %>% 
+  as.data.frame()
+
+hash_top <- 
+  hash_10_chr %>% 
+  slice_head(n = 4)
+
+hash_middle <- 
+  hash_10_chr %>% 
+  slice(1) %>% 
+  mutate(across(everything(), ~ ":"))
+
+rownames(hash_middle) <- ":"
+
+hash_bottom <- 
+  hash_10_chr %>% 
+  slice_tail(n = 4)
+
+bind_rows(hash_top, hash_middle, hash_bottom, hash_summary) %>%
+  rename(Agent = agent) %>%
+  mutate(Agent = remake_name(Agent)) %>% 
+  gt() %>% 
+  tab_spanner(label = "Features from Hashing", columns = c(-Agent)) %>% 
+  cols_label() %>%
+  # opt_row_striping() %>%
+  tab_style(style = cell_borders(sides = c("top", "bottom"),  weight = px(2)),
+            locations = cells_body(rows = 10)) %>%
+  tab_style(style = cell_fill(color = "white"),
+            locations = cells_body(rows = 10)) %>% 
+  cols_align(
+    align = "right",
+    columns = everything()
+  ) %>% 
+  tab_options(table.width = pct(70))
+
+```
+
+The main downside of this method is that the use of hash values makes it impossible to explain the model. If the tenth feature column is critical, we can't explain why this is the case for new data (since the hash function is practically non-reversible and may include collisions).  This may be fine if the primary objective is prediction rather than interpretation.  We also need to tune the number of hashing columns to optimize performance.
```

Comment:
```suggestion
The main downside of this method is that the use of hash values makes it impossible to explain the model. If the tenth feature column is critical, we can't explain why this is the case for new data (since the hash function is practically non-reversible and may include collisions).  This may be fine if the primary objective is prediction rather than interpretation.  When the goal is to optimize predictive performance, then the number of hashing columns to use can be included as a tuning parameter.  The model tuning process can then determine an optimal value of the number of hashing columns.
```